src/rvsdg/restructure.rs

```diff
@@ -306,93 +276,242 @@ impl SwitchCfgFunction {
-                        self.split_arc(e);
-                    } else if conts.len() > 1 {
-                        self.split_edges(src);
-                    }
+        if cont.reentry_nodes.is_empty() {
+            // Nothing to do.
+            return;
+        }
+        // assert!(
+        //     !cont.reentry_nodes.is_empty(),
+        //     "No reentry nodes found [start: {start:?}, entry: {entry:?}, exit: {exit:?}]"
+        // );
+
+        if cont.reentry_nodes.len() == 1 {
+            // There are multiple branches that all converge to a single "tail"
+            // node. This is _almost_ the structure that we want, but we need to
+            // add empty placeholder blocks and a few other things.
+            let tail = *cont.reentry_nodes.iter().next().unwrap();
+            for (edge, target) in self
+                .graph
+                .edges_directed(start, Direction::Outgoing)
+                .map(|x| (x.id(), x.target()))
+                // We make recursive calls to this method in this loop, so need
+                // to copy edge information out in this frame.
+                .collect::<Vec<_>>()
+            {
+                let reentry_edges = &cont.exit_arcs[&edge];
+                // This edge goes directly to the tail. Create an empty basic
+                // block (this will make our CFG look like a diamond rather than
+                // a triangle, which in turn makes it easier for us to generate
+                // a 'passthrough region' for this branch in the RVSDG).
+                if target == tail {
+                    self.split_arc(edge);
+                    continue;
                 }
-            }
 
-            // The rest of this loop is focused on mux/demux for multiple
-            // continuations. If we only have one, we are done.
-            if conts.len() == 1 {
-                continue;
+                if reentry_edges.len() == 1 {
+                    // We don't go directly to the tail, but there is at least a
+                    // single exit from the subgraph pointed to by `edge`.
+                    //
+                    // Restructure this subgraph (recursively).
+                    let next_edge = *reentry_edges.iter().next().unwrap();
+                    debug_assert_ne!(next_edge, edge);
+                    let branch_end = self.graph.edge_endpoints(next_edge).unwrap().0;
+                    self.restructure_branches_inner(target, branch_end, state);
+                    continue;
+                }
+                // There are multiple exit edges to the tail. Have them all join
+                // into an intermediate block.
+                let inter = self.fresh_block();
+                self.graph.add_edge(inter, tail, JMP);
+                for e in reentry_edges {
+                    self.divert_edge(*e, inter);
+                }
+                self.restructure_branches_inner(target, inter, state);
```

Comment:
In this recursive case, do we necessarily hit the if statement that came before this?
If so, can that be a helper that we call here?

Reply:
I don't think so. We don't actually know anything about target here, just that it isn't dominated by any of the outgoing edges. Unless I'm misunderstanding which `if` you're talking about.